[ClangImporter] Teach the importer about a couple more kinds macro constants

karwa opened the pull request:
[ClangImporter] Teach the importer about adding and subtracting integer constants, and to always look through macro definitions for infix operations.

<!-- What's in this pull request? -->

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

moiseev commented:
Is the `a.out` file necessary?

karwa commented:
Whoops, my bad. It's gone now.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
LGTM, but @jrose-apple or @milseman might have additional feedback.

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -424,118 +440,158 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
+    // (INT1 + INT2)
+    if (tokenI[1].is(clang::tok::plus)) {
+      llvm::APSInt value = firstValue + secondValue;
+      return createMacroConstant(impl, macro, name, DC, firstType,
```
I'm not sure it's correct to always use the first type for symmetric operations. Shifts can get away with it because the shift amount isn't really related to the quantity being shifted, but addition is a little sketchy. Maybe for now we should bail out if the types don't match.

jrose-apple replied:
Also, all the `createMacroConstant` calls are the same once we have a valid type and APSInt value. Maybe that can be factored out too?

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -424,118 +440,158 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
+    llvm::APSInt firstValue;
+    Type firstType = nullptr;
+
+    // Parse integer literal.
```
Can we factor this section out, since it's repeated twice?

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -424,118 +440,158 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
+    }
+    // AND, OR (bitwise, logical):
+    // (INT1 | INT2), (INT1 & INT2), (INT1 || INT2), (INT1 && INT2)
+    else if(isBinaryOperator(tokenI[1])) {
```
Nitpick: Missing space, and also we prefer to put `else if` on the same line as the close-brace.

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -424,118 +440,158 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
+    }
+    // Left-shift:
+    // (INT1 << INT2)
+    else if (tokenI[1].is(clang::tok::lessless)) {
```
If we're adding all these other things, does it make sense to support `>>` as well? I'm sure it's rare but even so.

jrose-apple commented on test/ClangImporter/macros.swift:
```diff
@@ -109,6 +109,20 @@ func testBitwiseOps() {
+  _ = ADD_TWO
+
+  _ = SUB_ONE
+  _ = SUB_ZERO
```
Hm. I'd really like a test to make sure the *values* are right too. I was hoping we had a SILGen test for this already but I don't see one. The closest I can see is ctypes_ir.swift, which checks for the values of imported enums.

Are you comfortable writing a test with -emit-silgen, or would you prefer I add one on your behalf after this gets merged?

jrose-apple reviewed (changes requested):
Looking good! Only one serious concern, and then a bunch of minor suggestions.

slavapestov commented:
@karwa Do you mind adding an entry to CHANGLEOG.md as well? We have several new ClangImporter features that might affect users so I'm following up with PR authors about documenting them.

karwa commented:
@jrose-apple I've added a bunch more operations, and changed the imported type of logical operations (&&, ||) to `Bool`.

Still working on the SILGen test, but I'd appreciate some feedback on mixed-sign arithmetic: I've tried to follow C's integer promotion, but I couldn't figure out how to get the bit-width of the signed value's type or how to look-up the stdlib integer type with the given bitwidth.

slavapestov commented:
@airspeedswift Might have some insights about mapping C integer types to Swift.

jrose-apple reviewed (commented):
I think you can write this using `clang::ASTContext::getIntTypeForBitwidth`. And to not duplicate promotion logic, there's already a clang::APSIntType (currently buried in the static analyzer) that can figure out the best type to promote to.

springsup committed:
[ClangImporter] Teach the importer about more infix operators between
integer constants, and to always look through macro definitions for them.

Also, logical comparisons now return a Boolean.

New operations: +, -, *, /, ^, >>, ==, >, >=, <, <=

springsup committed:
Update CHANGELOG

karwa commented:
@jrose-apple Addressed feedback. The promotion logic is really just to be explicit, anyway - the widest type always wins and if equal with mismatched sign, the signed becomes unsigned. If there are more nuanced ranking questions, they won't happen with our types anyway.

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -424,120 +459,186 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
+      } else if (secondValue.isSigned() && secondBitWidth > firstBitWidth) {
+        resultType = { secondValue, secondType };
+      } else {
+        llvm_unreachable("Types have unequal bitwidths, yet neither is greater "
```
What about when the first value is signed but the first bit width is smaller than the second?

jrose-apple replied:
Sorry, of course that's caught by the above, never mind.

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -424,120 +459,186 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
+    auto firstBitWidth  = firstValue.getBitWidth();
+    auto secondBitWidth = secondValue.getBitWidth();
+
+    // Ensure both values have the same signedness.
```
The logic is pretty simple, yes, but it's also just `std::max(APSIntType(firstValue), APSIntType(secondValue))`, and since I found one bug in it already I'm not sure it's worth risking another one.

karwa replied:
oh cool, I didn't spot that you could use it like that. Much nicer now, thanks!

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -424,120 +459,186 @@ static ValueDecl *importMacro(ClangImporter::Implementation &impl,
+    // Logical OR.
+    } else if (tokenI[1].is(clang::tok::pipepipe)) {
+      auto result  = firstValue.getBoolValue() || secondValue.getBoolValue();
+      resultValue  = result ? llvm::APSInt::get(1) : llvm::APSInt::get(0);
```
`llvm::APSInt::get(result)`?

jrose-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -7160,8 +7160,16 @@ ClangImporter::Implementation::createConstant(Identifier name, DeclContext *dc,
-                                              /*Implicit=*/true);
+      if (type->getCanonicalType()->isBool()) {
+        llvm::APSInt intValue = value.getInt();
+        assert((0 == intValue || 1 == intValue) && "Not a valid boolean");
```
Is this correct if someone has a macro value `(bool)5`?

karwa replied:
We don't support that, but sure, I guess we could in the future...

slavapestov commented:
@swift-ci Please smoke test

karwa commented:
@jrose-apple good to go?

jrose-apple commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -299,6 +293,47 @@ static Optional<clang::QualType> builtinTypeForToken(const clang::Token &tok,
+static Optional<std::pair<llvm::APSInt, Type>>
+  getIntegerConstantForMacroToken(ClangImporter::Implementation &impl,
+  	                              DeclContext *DC,
+    	                            const clang::Token &token) {
```
Some tabs crept in here instead of spaces. We would usually go with

```c++
static Optional<std::pair<llvm::APSInt, Type>>
getIntegerConstantForMacroToken(ClangImporter::Implementation &impl,
                                DeclContext *DC,
                                const clang::Token &token) {
```

jrose-apple reviewed (commented):
Just one last formatting change!

jrose-apple commented:
@swift-ci Please test

IngmarStein commented:
Before this PR, the following macro from `/usr/include/mach/machine.h` had been imported successfully but is now invisible to Swift code (just an example, there are others):
```
#define	CPU_TYPE_X86_64		(CPU_TYPE_X86 | CPU_ARCH_ABI64)
```
The individual components `CPU_TYPE_X86` and `CPU_ARCH_ABI64` are imported, though.

jrose-apple commented:
Hm. I'm not seeing that. Both my Swift 3.1 and my recently-built Swift (from last week) give me `$R0: cpu_type_t = 16777223`. Can you file a bug at https://bugs.swift.org with repro steps?

IngmarStein commented:
Inded, the error seems to be in another Xcode component (I tried the Xcode 9 beta). I see `Use of unresolved identifier 'CPU_TYPE_POWERPC64'. Did you mean 'CPU_TYPE_POWERPC'` only when Xcode automatically compiles the file, but the error disappears when doing a full build and cannot be reproduced in the REPL. SourceKit then?

jrose-apple commented:
Possibly. Or possibly an issue with precompiled bridging headers. Either way, bug please!

IngmarStein commented:
Sure, here you go: https://bugs.swift.org/browse/SR-5113. Sorry for commenting in this closed PR.

karwa commented:
No apology necessary - it's clear why you might think this PR was to blame (and it still might be; although if it works in a plain, non-Xcode toolchain, it sounds like the problem might originate elsewhere)
